markdown/issues.md

```diff
@@ -1,10 +1,82 @@
+
+Although we had a disaster recovery setup in another region, we couldn't initiate failover because the Kubernetes API services, load-balanced by Octavia, were inaccessible. As a result, we were unable to access the cluster or execute any commands, hindering the Primary/DR switch that required running scripts against the Kubernetes cluster. This incident exposed our lack of preparedness for such an outage, highlighting flaws in our failover process. None of our systems will function until Octavia is fixed.
+
+Unfortunately, Octavia is always addressed last, as its functionality depends on the proper functioning of Nova and Neutron. Therefore, after a complete site failure, the controllers, Nova and Neutron, must be restored before investigating and fixing Octavia and the broken load balancers.
+
+Even after these issues are solved and we can communicate  with the Octavia API, often the loadbalancers are stuck in an ERROR state, which we can't fix it on 
+our own and need admin privileges to fix.
+
+To mitigate such downtime during outages, we explored an alternative solution to expose our Kubernetes service without relying on Octavia.
+We use Caddy as a web server as it is used in Tutor both as a web proxy and for generating SSL/TLS certificates at runtime.
+
+We were looking for a solution to expose the caddy service other than the loadbalancer and make our platform Octavia free. According to the Kubernetes documentation, several ways exist to expose a service in Kubernetes.
+
+* The first approach, ClusterIP, allows the service to be accessed only via an internal IP within the cluster. However, this type of exposure is not suitable for a production environment as it restricts accessibility within the cluster.
+
+* The second option, NodePort, is also unsuitable for our needs since our Caddy service operates on ports 80 and 443, while the NodePort range is from 30000 to 32767.
+
+* The third option is a loadbalancer, which we are trying to eliminate.
+* Lastly, there is a way not widely used among the kubernetes community because many are using cloud provider's Load balancers.
+Using ClusterIP with external IP
+
+Customers can expose their Kubernetes cluster to external traffic using ClusterIP and an external IP address. The Kubernetes service type is ClusterIP, but an external IP address can be specified using the externalIP property. This solution requires that the customer maps the external IP address of a service to their private IP addresses in the cluster. 
+
+Unfortunately, we encountered issues in utilizing a Floating IP address as an external IP address. It was due to the fact that our Floating IP address was mapped to a private address within a different subnet than the one used by the ClusterIP.
+
+So, we were not able to figure out any way to expose our services without Loadbalancer.
```

Comment:
Well, no we weren't, but we _were_ able to come up with an approach where rather than doing anything with the Kubernetes APIs, we would just operate on a single DNS record that we could flip, without using any API calls, to the other site. That allows us to have at least a rudimentary service on the backup site, until load balancer functionality in the failed primary site becomes available again.